torchrl/objectives/dqn.py

```diff
@@ -95,8 +96,15 @@ def forward(self, input_tensordict: TensorDictBase) -> TensorDict:
             buffers=self.value_network_buffers,
         )
 
+        action = tensordict.get("action")
         pred_val = td_copy.get("action_value")
-        pred_val_index = (pred_val * action).sum(-1)
+
+        if _CATEGORICAL_ACTION_ENCODING:
+            batch_size = action.size(0)
+            pred_val_index = pred_val[range(batch_size), action.squeeze(-1)]
```

Comment:
I guess we'd like to be able to work with batch sizes that are more than unidimensional.

We can use `gather` (I think)
```python
x = torch.randn(3, 4, 5)
idx = torch.randint(5, (3, 4))
x.gather(-1, idx)
```

This will index `x` along the last dimension.

Reply:
this is not covered by the tests (it should be done in `test_costs.py` I believe).

Reply:
Reworked indexing here, thx

Reply:
Also added tests so should be covered by coverage by now